tree/ntuple/v7/inc/ROOT/RCluster.hxx

```diff
@@ -0,0 +1,192 @@
+         return lhs.fColumnId == rhs.fColumnId && lhs.fPageNo == rhs.fPageNo;
+      }
+   };
+
+   ROnDiskPage() = default;
+   ROnDiskPage(void *address, std::size_t size) : fAddress(address), fSize(size) {}
+
+   const void *GetAddress() const { return fAddress; }
+   std::size_t GetSize() const { return fSize; }
+
+   bool IsNull() const { return fAddress == nullptr; }
+};
+
+} // namespace Detail
+} // namespace Experimental
+} // namespace ROOT
+
+// For hash maps ROnDiskPage::Key --> ROnDiskPage
+namespace std
+{
+   template <>
+   struct hash<ROOT::Experimental::Detail::ROnDiskPage::Key>
+   {
+      // TODO(jblomer): quick and dirty hash, likely very sub-optimal, to be revised later.
+      size_t operator()(const ROOT::Experimental::Detail::ROnDiskPage::Key &key) const
+      {
+         return ((std::hash<ROOT::Experimental::DescriptorId_t>()(key.fColumnId) ^
+                 (hash<ROOT::Experimental::NTupleSize_t>()(key.fPageNo) << 1)) >> 1);
+      }
+   };
+}
+
+
+namespace ROOT {
+namespace Experimental {
+namespace Detail {
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::ROnDiskPageMap
+\ingroup NTuple
+\brief A memory region that contains packed and compressed pages
+
+Derived classes implement how the on-disk pages are stored in memory, e.g. mmap'd or in a special area.
+*/
+// clang-format on
+class ROnDiskPageMap {
+   friend class RCluster;
+
+protected:
+   /// The memory region containing the on-disk pages. Ownership of the memory region is passed to the page map.
+   /// Therefore, the region needs to be allocated in a way that fits the derived class and its destructor.
+   void *fMemory;
+   std::unordered_map<ROnDiskPage::Key, ROnDiskPage> fOnDiskPages;
+
+public:
+   explicit ROnDiskPageMap(void *memory) : fMemory(memory) {}
+   ROnDiskPageMap(const ROnDiskPageMap &other) = delete;
+   /// The default move constructor does not reset fMemory
+   ROnDiskPageMap(ROnDiskPageMap &&other);
+   ROnDiskPageMap &operator =(const ROnDiskPageMap &other) = delete;
+   /// The default move assignment does not reset fMemory
+   ROnDiskPageMap &operator =(ROnDiskPageMap &&other);
+   virtual ~ROnDiskPageMap();
+
+   /// Inserts information about a page stored in fMemory.  Therefore, the address referenced by onDiskPage
+   /// needs to be owned by the fMemory block.  If a page map contains a page of a given column, it is expected
+   /// that _all_ the pages of that column in that cluster are part of the page map.
+   void Register(const ROnDiskPage::Key &key, const ROnDiskPage &onDiskPage) { fOnDiskPages.emplace(key, onDiskPage); }
+};
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::ROnDiskPageMapHeap
+\ingroup NTuple
+\brief An ROnDiskPageMap that is used for an fMemory allocated as an array of unsigned char.
+*/
+// clang-format on
+class ROnDiskPageMapHeap : public ROnDiskPageMap {
+public:
+   explicit ROnDiskPageMapHeap(void *memory) : ROnDiskPageMap(memory) {}
```

Comment:
As the dtor destroys that memory, signaling that through a `unique_ptr<char[]>` sounds a lot nicer. Is any ROnDiskPageMap owning the memory, or just `ROnDiskPageMapHeap`? If the former, the base could have a member of type `unique_ptr`. If the latter, consider moving `fMemory` to the derived classes and using `unique_ptr<char[]>` where the memory is owned.

Reply:
The `fMemory` area is now an `std::unique_ptr<unsigned char []>` in `ROnDiskPageMapHeap`.